[LLDB] Add more helper functions to ValueObject class.

cmtice opened the pull request:
Create additional helper functions for the ValueObject class, for:
  - returning the value as an APSInt or APFloat
  - additional type casting options
  - additional ways to create ValueObjects from various types of data
  - dereferencing a ValueObject

These helper functions are needed for implementing the Data Inspection Language, described in
https://discourse.llvm.org/t/rfc-data-inspection-language/69893

cmtice committed:
[LLDB] Add more helper functions to ValueObject class.

Create additional helper functions for the ValueObject class, for:
  - returning the value as an APSInt or APFloat
  - additional type casting options
  - additional ways to create ValueObjects from various types of data
  - dereferencing a ValueObject

These helper functions are needed for implementing the Data Inspection
Language, described in
https://discourse.llvm.org/t/rfc-data-inspection-language/69893

jimingham commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -441,6 +441,19 @@ class ValueObject {
 
   virtual int64_t GetValueAsSigned(int64_t fail_value, bool *success = nullptr);
 
+  llvm::APSInt GetValueAsAPSInt();
```
Not all SBValues are representable as Int's or Floats, Bool's etc.  But I don't see any way to report errors with these new API's.

adrian-prantl replied:
if these errors matter to users then `llvm::Expected<llvm::APSInt>` would be the right choice. Otherwise std::optional is the way to go.

jimingham commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -618,6 +631,24 @@ class ValueObject {
+
+  lldb::ValueObjectSP CastIntegerOrEnumToEnumType(CompilerType type);
+
+  lldb::ValueObjectSP CastFloatToEnumType(CompilerType type, Status &error);
```
You don't need an error parameter for anything that returns a ValueObject.  ValueObjects carry their errors in the ValueObject (ValueObject::GetError()).

jimingham commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
   CreateValueObjectFromData(llvm::StringRef name, const DataExtractor &data,
                             const ExecutionContext &exe_ctx, CompilerType type);
 
+  static lldb::ValueObjectSP
```
Given targets can be big or little-endian, I'm a little worried about a "const void *bytes" that says nothing about what is in the bytes.  This is what DataBuffer's are for, why couldn't you use them?

clayborg replied:
I agree  on this, but why can't we just use the above `CreateValueObjectFromData(llvm::StringRef name, const DataExtractor &data, const ExecutionContext &exe_ctx, CompilerType type);`? The `DataExtractor` in the mentioned existing API can own or share a pointer to data. Would the "const void *bytes" always be copied into the internal data in the `ValueObject` in this case?

cmtice replied:
CreateValueObjectFromBytes was being called from CreateValueObjectfrom{APInt,Pointer,Bool,Nullptr}, receivng the pointer to the appropriate values in the input parameter, putting the bytes into a DataExtractor, then calling CreateValueObjectFromData with the DataExtractor. I thought that would handle things like correct endianness, etc.  However, I will move the "putting the bytes into the DataExtractor" into the callers for CreateValueObjectFromBytes & call CreateValueObjectFromData directly from each of them. This leads to a bit of code duplication, but I can see that it's safer. (This eliminates the CreateValueObjectFromBytes functions altogether).

adrian-prantl commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -441,6 +441,19 @@ class ValueObject {
+
+  llvm::APFloat GetValueAsFloat();
+
+  bool GetValueAsBool();
```
same here

adrian-prantl commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+  CreateValueObjectFromBytes(lldb::TargetSP target_sp, const void *bytes,
+                             CompilerType type);
+
+  static lldb::ValueObjectSP CreateValueObjectFromBytes(lldb::TargetSP target,
```
These functions could all benefit from Doxygen comments. If we can use a StringRef, or DataExtractor/Buffer instead a raw pointer, that would likely also be safer.

bulbazord replied:
+1

jimingham commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+  CreateValueObjectFromAPFloat(lldb::TargetSP target, const llvm::APFloat &v,
+                               CompilerType type);
+
+  static lldb::ValueObjectSP CreateValueObjectFromPointer(lldb::TargetSP target,
```
How is this different from CreateValueObjectFromAddress?

clayborg replied:
I agree, we already have this API unless this is doing things differently. That other API allows a name to be specified and is used for synthetic children, so you should be able to use it and specify no name

cmtice replied:
"CreateValueObjectFromPointer" is different from "CreateValueObjectFromAddresss" In that the former creates a value object whose value is the pointer address; the latter creates a value object whose value is whatever the pointer points to (it dereferences the pointer).  I *can* update CreateValueObjectFromAddress to take a parameter telling it whether to dereference the pointer or use the pointer address as the value, if you would prefer.

jimingham commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -719,6 +776,10 @@ class ValueObject {
     ClearUserVisibleData(eClearUserVisibleDataItemsSummary);
   }
 
+  void SetDerefValobj(ValueObject *deref) { m_deref_valobj = deref; }
```
Synthetic child providers also have a way to provide the deref ValueObject dynamically.  How do these two ways of doing that job work together?

cmtice replied:
I use the synthetic child providers way of getting the deref ValueObject  to get the value I use for my call to SetDerefValobj (when I'm converting a ValueObject containing a SmartPointer into a ValueObject containing the pointer the SmartPointer was referencing).

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+
+  for (const uint32_t i : idx) {
+    // Force static value, otherwise we can end up with the "real" type.
+    inner_value = GetChildAtIndex(i, /*can_create_synthetic*/ false);
```
What does this do exactly? The result of the loop is that `inner_value` is pointing to the last child, but I'm struggling to understand the intent here

cmtice replied:
Ah, OOPS! I just re-discovered this comment, and it's actually a bug in my code (I mis-copied something). It should be "inner_value = inner_value.GetChildAtIndex(...). I will fix that and send the fix later today.

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+  CompilerType inner_value_type = inner_value->GetCompilerType();
+  if (type.IsPointerType()) {
+    assert(inner_value_type.CompareTypes(type.GetPointeeType()) &&
+           "casted value doesn't match the desired type");
```
Same with this assertion.

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+
+  // At this point the target type should be a reference.
+  assert(inner_value_type.CompareTypes(type.GetNonReferenceType()) &&
+         "casted value doesn't match the desired type");
```
Same w/ this assertion.

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+  lldb::TargetSP target = GetTargetSP();
+
+  assert((type.IsPointerType() || type.IsReferenceType()) &&
+         "invalid ast: target type should be a pointer or a reference");
```
Same w/ this assertion

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+lldb::ValueObjectSP ValueObject::CastScalarToBasicType(CompilerType type,
+                                                       Status &error) {
+  assert(type.IsScalarType() && "target type must be an scalar");
+  assert(GetCompilerType().IsScalarType() && "argument must be a scalar");
```
Same for these assertions

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+      return ValueObject::CreateValueObjectFromAPFloat(target, f, type);
+    }
+  }
+  assert(false && "invalid target type: must be a scalar");
```
Looks like this was leftover as a debug assertion?

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+  lldb::TargetSP target = GetTargetSP();
+
+  assert(type.IsScalarType() && "target type must be a scalar");
+  assert(GetCompilerType().IsEnumerationType() && "argument must be an enum");
```
Same w/ these assertions.

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+        type.GetCanonicalType().GetBasicTypeEnumeration());
+    return ValueObject::CreateValueObjectFromAPFloat(target, f, type);
+  }
+  assert(false && "invalid target type: must be a scalar");
```
Debug assertion?

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+    val_byte_size = temp.value();
+  assert(type.IsInteger() && "target type must be an integer");
+  assert((type.IsBoolean() || type_byte_size >= val_byte_size) &&
+         "target type cannot be smaller than the pointer type");
```
Same w/ these assertions

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+  assert(type.IsEnumerationType() && "target type must be an enum");
+  assert((GetCompilerType().IsInteger() ||
+          GetCompilerType().IsEnumerationType()) &&
+         "argument must be an integer or an enum");
```
Same w/ these assertions

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+  lldb::TargetSP target = GetTargetSP();
+
+  assert(type.IsEnumerationType() && "target type must be an enum");
+  assert(GetCompilerType().IsFloat() && "argument must be a float");
```
Same w/ these assertions

cmtice replied:
Whoops! Good catch. I've fixed that now.

bulbazord commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -3031,6 +3378,64 @@ lldb::ValueObjectSP ValueObject::CreateValueObjectFromData(
+lldb::ValueObjectSP
+ValueObject::CreateValueObjectFromNullptr(lldb::TargetSP target,
+                                          CompilerType type) {
+  assert(type.IsNullPtrType() && "target type must be nullptr");
```
Same for this assertion

bulbazord commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -618,6 +631,24 @@ class ValueObject {
+
+  lldb::ValueObjectSP CastEnumToBasicType(CompilerType type);
+
+  lldb::ValueObjectSP CastPointerToBasicType(CompilerType type);
```
What is `BasicType` in this context? I know there's `lldb::BasicType`, but these methods take `CompilerType`s.

bulbazord reviewed (changes requested):
+1 to all of Jim and Adrian's comments.

High level comments:
1. I'm a little concerned about the use of assertions in this patch. I generally like assertions when they are used to assert internal consistency. It looks like you're using them as an error-checking and parameter-checking mechanism. This seems pretty fragile, especially since it will be difficult to enforce that these functions are called "correctly" in all cases from all contributors.

2. There isn't a lot of documentation. It would be helpful to know the expected behavior for a lot of these methods (especially for people like me  who aren't super familiar with the ValueObject class and all of its tendrils).

3. Is there a way you can add some tests for the new functionality? I'm not sure how you could trigger these new methods from an API test, but a unit test might be possible?

clayborg commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+  CreateValueObjectFromBytes(lldb::TargetSP target_sp, const void *bytes,
+                             CompilerType type);
+
+  static lldb::ValueObjectSP CreateValueObjectFromBytes(lldb::TargetSP target,
```
Do we need to be able to add a name for the value object? Some the creation APIs allow users to specify names and the native synthetic child providers could use these APIs if they can specify a name.

clayborg commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+                                                        const void *bytes,
+                                                        lldb::BasicType type);
+
+  static lldb::ValueObjectSP CreateValueObjectFromAPInt(lldb::TargetSP target,
```
ditto for adding a name parameter.

clayborg commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+                                                        CompilerType type);
+
+  static lldb::ValueObjectSP
+  CreateValueObjectFromAPFloat(lldb::TargetSP target, const llvm::APFloat &v,
```
ditto for adding a name parameter.

cmtice replied:
Done

clayborg commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+                                                          uintptr_t addr,
+                                                          CompilerType type);
+
+  static lldb::ValueObjectSP CreateValueObjectFromBool(lldb::TargetSP target,
```
name arg?

cmtice replied:
Done

clayborg commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -668,6 +699,32 @@ class ValueObject {
+  static lldb::ValueObjectSP CreateValueObjectFromBool(lldb::TargetSP target,
+                                                       bool value);
+
+  static lldb::ValueObjectSP CreateValueObjectFromNullptr(lldb::TargetSP target,
```
name arg?

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+  if (auto temp = GetCompilerType().GetByteSize(target.get()))
+    byte_size = temp.value();
+
+  unsigned bit_width = static_cast<unsigned>(byte_size * CHAR_BIT);
+  bool success = true;
+  uint64_t fail_value = 0;
+  uint64_t ret_val = GetValueAsUnsigned(fail_value, &success);
+  uint64_t new_value = fail_value;
+  if (success)
+    new_value = ret_val;
+  bool is_signed = GetCompilerType().IsSigned();
```
Extract the `APInt` from the contained `Scalar` value directly. No need to create a location uint64_t and then try to re-create, the `Scalar` ivar already contains an `APInt`. So just do:
```
  if (CanProvideValue()) {
    Scalar scalar;
    if (ResolveValue(scalar)) {
      if (success)
        *success = true;
      scalar.MakeUnsigned();
      return scalar.m_integer; // Might need to provide an accessor for Scalar::m_integer
    }
  }
```

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+  return llvm::APSInt(llvm::APInt(bit_width, new_value, is_signed), !is_signed);
+}
+
+llvm::APFloat ValueObject::GetValueAsFloat() {
```
Resolving the float value already builds a `APFloat` in a `Scalar`, so no need to redo all of this. Should be very similar to `llvm::APSInt ValueObject::GetValueAsAPSInt()` suggestions above

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+
+  switch (basic_type) {
+  case lldb::eBasicTypeFloat: {
+    float v = 0;
```
We really don't want to use native float types, we want to use real APFloat values that are correctly encoded. For instance if we are debugging from x86_64, we don't want to create a 10 byte Intel float and do the math with the native float, we want to use the value from the Scalar that is already set correctly. If we are cross debugging to an iPhone, we want the floating point values to match exactly, which requires us to use the APFloat and not do any manually create a native float and convert it. This all works for the value objects when they get the float as a string, and will be similar where we resolve the scalar and then can just use the Scalar::m_float directly.

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+
+bool ValueObject::GetValueAsBool() {
+  CompilerType val_type = GetCompilerType();
+  if (val_type.IsInteger() || val_type.IsUnscopedEnumerationType() ||
+      val_type.IsPointerType()) {
+    return GetValueAsAPSInt().getBoolValue();
+  }
+  if (val_type.IsFloat()) {
+    return GetValueAsFloat().isNonZero();
+  }
```
It might be better to just resolve the scalar like suggested for `llvm::APSInt ValueObject::GetValueAsAPSInt()` and if that succeeds, then check the Scalar type for `Scalar::e_int` or `Scalar::e_float` and then return the right value from there. I am not sure that the `if` checks above are going to cover all types we want to be able to get an boolean value for.

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+  if (val_type.IsFloat()) {
+    return GetValueAsFloat().isNonZero();
+  }
+  if (val_type.IsArrayType()) {
+    lldb::ValueObjectSP new_val =
+        ValueObject::ValueObject::CreateValueObjectFromAddress(
+            GetName().GetStringRef(), GetAddressOf(), GetExecutionContextRef(),
+            val_type);
+    return new_val->GetValueAsUnsigned(0) != 0;
+  }
```
What is this attempting to do? It looks like this is taking an array type and returning an array pointer and trying to return `true` if the array is in memory?

cmtice replied:
Hm...What I was trying to do (and I can see I didn't do it well here) is to check to see if the array had a real value(pointer address) or not.  I'm updating the code to be cleaner and more clear.

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+  if (auto temp = GetCompilerType().GetByteSize(target.get()))
+    byte_size = temp.value();
+
+  assert(value.getBitWidth() == byte_size * CHAR_BIT &&
```
yes, no asserts and add error propagation.

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+  return false;
+}
+
+void ValueObject::UpdateIntegerValue(const llvm::APInt &value) {
```
If I look at this API, I would expect this function to change the integer value of any ValueObject. The value object might have a variable that lives in memory (on the stack or heap), or if it lives in a register (update the register) or if there is already a local buffer, replace the bytes in the local buffer. 

This function seems to replace the integer value always with a local buffer?

cmtice replied:
The intent is to only call this function on ValueObjects that are intermediate objects in the new parser/evaluator, not ValueObjects that are associated with program variables. I'll add a check for that while I'm updating the other asserts & error checks.  So no, this function should not update values in memory, heap or register, only in the ValueObject itself.

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+  assert(value.getBitWidth() == byte_size * CHAR_BIT &&
+         "illegal argument: new value should be of the same size");
+
+  lldb::DataExtractorSP data_sp;
+  Status error;
+  data_sp->SetData(value.getRawData(), byte_size,
+                   target->GetArchitecture().GetByteOrder());
```
This will crash, `data_sp` is never constructed or assigned. We also need to figure out if we are going to want the integer value to be replaced always with a local copy, or updated in the actual program memory, register, or local buffer.

I realize many of these APIs are probably only intended to be used in the new ValueObject based expression parser, but there are APIs being added to ValueObject so they should work regardless of where they come from or originate.

cmtice replied:
Just for the record: This call is not crashing for me; I am calling this code in my light-weight parser/evaluator, and it seems to work just fine. 

I am ok with changing this code, if you really think it should be changed, but I'd appreciate suggestions on how to make this work "properly" -- I've been looking through ValueObject.cpp & ValueObject.h and not seeing much that looks helpful to me. This does seem to be the way SBValue::SetData works (which was the original basis for this function...)

clayborg commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2919,243 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+        module_sp->ResolveFileAddress(addr_value, tmp_addr);
+        addr_value = tmp_addr.GetLoadAddress(target_sp.get());
+      }
+    } else if (addr_type == eAddressTypeHost || addr_type == eAddressTypeHost)
```
Double check for `eAddressTypeHost`, I am guessing the second one was meant to be `eAddressTypeInvalid`. We should also change `lldb::addr_t SBValue::GetLoadAddress()` to call this function

cmtice commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -441,6 +441,19 @@ class ValueObject {
+
+  llvm::APFloat GetValueAsFloat();
+
+  bool GetValueAsBool();
```
I would prefer to have GetValueAsBool().  Otherwise I will need to do integer-to-bool conversions in the places where I call this function; also it makes the code more clear, readable and maintainable.  But if you feel very strongly about this, I can fall back on GetValueAsAPSInt.

cmtice commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -441,6 +441,19 @@ class ValueObject {
+  /// parameter.
+  void UpdateIntegerValue(const llvm::APInt &value);
+
+  /// Assign the integer value 'new_val_sp' to the current object.
+  void UpdateIntegerValue(lldb::ValueObjectSP new_val_sp);
```
Yes, this assigns the value from new_val_sp to the current (integer) ValueObject.  The main part of this function does call the preceding one. Because new_val_sp can actually contain an integer, a float or a pointer, there's some slightly different manipulations of the value that needs to happen before calling the preceding function (take a look at the code in ValueObject.cpp), which is why this function was written.

cmtice commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -618,6 +631,24 @@ class ValueObject {
                                               lldb::TypeSP &type_sp);
 
+  /// Return the target load address assocaited with this value object.
+  lldb::addr_t GetLoadAddress();
```
Yes, this is based on SBValue::GetLoadAddress; I'll update that one to call this one.  As written, it already does return LLDB_INVALID_ADDRESS in certain cases...I'm not sure what you want me to do differently (probably better to discuss this at the implementation rather than the declaration).

jimingham commented:
How does this do error handling?  If I pass in new_val_sp which contains a "struct Foo" - so really not convertible to an integer, this will fail.  Presumably when that happens, you just back out of setting the original ValueObject's value?  So it doesn't seem appropriate to use that ValueObject's Status object to report the error.  And it wasn't new_value_sp's fault you used it incorrectly, so you also can't change its Status.

So I think in this case you do need to pass in an Status object.

In general, it seems like these additions have not thought very much about error reporting...

Jim


> On Apr 2, 2024, at 8:38 PM, cmtice ***@***.***> wrote:
> 
> 
> @cmtice commented on this pull request.
> 
> In lldb/include/lldb/Core/ValueObject.h <https://github.com/llvm/llvm-project/pull/87197#discussion_r1548868655>:
> 
> > +  /// Assign the integer value 'new_val_sp' to the current object.
> +  void UpdateIntegerValue(lldb::ValueObjectSP new_val_sp);
> Yes, this assigns the value from new_val_sp to the current (integer) ValueObject. The main part of this function does call the preceding one. Because new_val_sp can actually contain an integer, a float or a pointer, there's some slightly different manipulations of the value that needs to happen before calling the preceding function (take a look at the code in ValueObject.cpp), which is why this function was written.
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/llvm/llvm-project/pull/87197#discussion_r1548868655>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/ADUPVW5CJAYIC6EDCGLDVYTY3N2RZAVCNFSM6AAAAABFQTAPTWVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMYTSNZVGQZDEMBRHE>.
> You are receiving this because you are on a team that was mentioned.
>

cmtice commented:
Most of my error checking was being done in the code that calls these functions, rather than in the functions directly, but I see your point. I will update the functions to do the error checking & reporting directly themselves.

cmtice commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+    new_value = ret_val;
+  bool is_signed = GetCompilerType().IsSigned();
+
+  return llvm::APSInt(llvm::APInt(bit_width, new_value, is_signed), !is_signed);
```
The call to GetValueAsUnsigned included a boolean, 'success' which would be set correctly if the call fails, and which I was checking upon return. But, if I use Greg's code from above then I omit the call to GetValueAsUnsigned, anyway.  And yes, I am converting these GetValueAs... functions to return an llvm::Expected<> value.

cmtice commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+    return llvm::APFloat(v);
+  }
+  default:
+    return llvm::APFloat(NAN);
```
I think it was coming from a stdlib definition; but I'm updating the code to set an error message here instead.

cmtice committed:
[LLDB] Add more helper functions to ValueObject class.

Fix various issues in  GetValueAsAPSInt, GetValueAsAPFloat,
GetValueAsBool, UpdateIntegerValue (renamed to SetValueFromInteger),
and GetLoadAddress:

- Added error checking & reporting.
- Removed asserts.
- Added doxygen comments.
- Renamed UpdateIntegerValue to SetValueFromInteger
- Renamed GetValueAsFloat to GetValueAsAPFloat
- Updated code in GetValue... functions to use existing Scalar values
  where possibled.
- Updated SBValue::GetLoadAddress to call ValueObject::GetLoadAddress.

cmtice committed:
Merge branch 'main' of github.com:llvm/llvm-project into dil-valobj-2

cmtice commented:
I think I've addressed all the review comments for GetValueAsAPSInt, GetValueAsAPFloat, GetValueAsBool, UpdateIntegerValue (not SetValueFromInteger) and GetLoadAddress.  I apologize if I missed anything. Please take another look at those functions.

I'm still working on fixing the issues with the other functions. :-)

cmtice committed:
[LLDB] Fix review comments for "CreateValueObjectFrom..." functions.

- Update CreateValueObjectFromAddress to optionally not dereference the address
  before creating the value (i.e. create a value object containing the addres
  itself). This removes the need for CreateValueObjectFromPointer.
- Remove CreateValueObjectFromPointer
- Remove CreateValueObjectFromBytes functions, moving the DataExtractor into
  the callers & having them call CreateValueObjectFromData directly.
- Add 'name' parameter to the next CreateValueObjectFrom... functions.
- Remove assert stetement.
- Add Doxygen comments.

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-valobj-2

cmtice commented:
I've now addressed the comments for the "CreateValueObjectFrom..." functions (I think).

I will not be able to work on this for the next couple of weeks, but I'll get back to it and fix the rest as soon as I can.

cmtice committed:
Merge remote-tracking branch 'origin/main' into dil-valobj-2

cmtice committed:
[LLDB] Add more helper functions to ValueObject class.

Add doxygen comments for cast functions; reduce number of cast functions
by combining some; fix error reporting in cast functions and remove all
assertions.

cmtice commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,116 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+          new_val_type.IsPointerType()) &&
+         "illegal argument: new value should be of the same size");
+
+  if (new_val_type.IsInteger()) {
```
Done (I thiink!)

cmtice commented:
I'm sorry this has taken me so long, but I believe I have addressed/fixed everyone's comments now; if I missed any please let me know.  Please review this PR again.

cmtice committed:
[LLDB] Add more helper functions to ValueObject.

Fix remaining issues wit 'cast' functions:
- Update parameter type checking
- Update error messages & error status checking
- Correctly update 'inner_value' in 'CastDerivedToBaseType'

cmtice commented:
I've updated with a few minor fixes. I've addressed all the review comments (I think).  I've verified/tested that all the casting, etc. works properly. Please review/approve this. Thanks!

jimingham commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -3015,11 +3017,12 @@ llvm::Expected<lldb::ValueObjectSP> ValueObject::CastDerivedToBaseType(
 
   for (const uint32_t i : base_type_indices)
     // Force static value, otherwise we can end up with the "real" type.
-    inner_value = GetChildAtIndex(i, /*can_create_synthetic*/ false);
```
I don't understand this change.  How is getting `this->GetSP()->GetChildAtIndex` different from `this->GetChildAtIndex`?

The other change here seems to be to change can_create_synthetic from false to true.  IIUC, the comment directly above this change says why it chose `false` here, though it is a somewhat confusing comment.  Since you seem to disagree with that comment, you should probably change the comment to say why you think `true` is the right value here.

cmtice replied:
The difference is that now in each loop iteration the rhs is 'inner_loop->GetChildAtIndex...', which will be different each iteration through the loop; before it was just 'GetChildAtIndex', which did not change on each iteration.

re false->true: The original lldb-eval code on which that was based used 'false', but it was also making the call through the SB API interface (SBValues, SBTypes, etc). I found that the cast test cases passed with lldb-eval when this value was false, but failed in my translated, migrated version. Setting this to true allows those failing test cases to pass.

I don't have any easy way to add those test cases to this PR because they all use 'frame variable' (using my DIL implementation) for doing the casts...But I do have them, and I am running (and passing) them.

jimingham replied:
Ah, I missed that I was looking at a diff against your original diff which didn't make sense.  In context the code seems fine.   

But then that also showed that you are the author of the "somewhat confusing comment".  Can you make that comment more helpful?  Suppressing the synthetic child production would force this child to get its type directly from the CompilerType, which I would call the "real" child.  So that comment seems to make sense when we passed `true` but doesn't make sense when passing true.

You don't need to introduce tests for this at this stage if that's dependent on other changes, but this seems like a tricky point, so it should have a comment.  But this one doesn't make sense to me.

cmtice replied:
Done.

jimingham commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -3138,13 +3141,13 @@ lldb::ValueObjectSP ValueObject::CastToBasicType(CompilerType type) {
 
   if (is_pointer) {
-    if (type.IsBoolean() && type_byte_size < val_byte_size) {
+    if (!type.IsBoolean() && type_byte_size < val_byte_size) {
```
The `!` here seems right, but still this logic is odd.  

With this code, if I pass 32 bit float (on a 64 bit pointers system), the error I'll get is that it's too small.  OTOH, if I pass a 64 bit float, I'll get "must be integer".  That's confusing.  If you aren't planning to take anything but integers, it seems like you should check that first, then check the size.

cmtice committed:
Fix comment.

cmtice committed:
[LLDB] Add more helper functions to ValueObject class.

Re-order checks for input type and size in CastToBasicType.

cmtice commented on lldb/include/lldb/Core/ValueObject.h:
```diff
@@ -618,6 +631,24 @@ class ValueObject {
   virtual lldb::ValueObjectSP CastPointerType(const char *name,
                                               lldb::TypeSP &type_sp);
 
+  /// Return the target load address assocaited with this value object.
```
Done

jimingham commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -3141,15 +3141,15 @@ lldb::ValueObjectSP ValueObject::CastToBasicType(CompilerType type) {
+      m_error.SetErrorString("target type must be an integer");
+      return GetSP();
+    }
     if (!type.IsBoolean() && type_byte_size < val_byte_size) {
```
That's better, but now the `type.IsBoolean` is entirely pointless, since you would never get to this code in that case.

cmtice replied:
Whoops! Good catch; I've fixed that now.

cmtice committed:
[LLDB] Add more helper functions to ValueObject class.

Fix type-checking error introduced with previous re-ordering fix.

jimingham commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -1089,6 +1089,154 @@ int64_t ValueObject::GetValueAsSigned(int64_t fail_value, bool *success) {
+      uint64_t num_bits = 0;
+      if (auto temp = new_val_sp->GetCompilerType().GetBitSize(target.get()))
+        num_bits = temp.value();
+      // SetValueFromInteger(llvm::APInt(64, int_val), error);
```
Is this leftover from some experiment?

cmtice replied:
Oh, yes. I thought I had found and removed all of those, but it looks like I missed one. I'll remove it.

cmtice replied:
Done.

jimingham commented:
There was one comment that should probably be removed.  This looks okay to me at this stage.

cmtice committed:
[LLDB] Add more helper functions to ValueObject class.

Delete a commented-out line of code.

cmtice commented:
> There was one comment that should probably be removed. This looks okay to me at this stage.

Jim, If this looks ok to you, could you officially mark it as approved please?

cmtice commented:
Review ping? Does this look ok to everyone now? I would really like to get this PR committed (I have more waiting behind it). Thanks in advance!

jimingham reviewed (approved):
LGTM at this point.

adrian-prantl commented on lldb/source/Core/ValueObject.cpp:
```diff
@@ -2809,6 +2956,374 @@ ValueObjectSP ValueObject::CastPointerType(const char *name, TypeSP &type_sp) {
+
+  if (!type.IsScalarType()) {
+    m_error.SetErrorString("target type must be a scalar");
+    return GetSP();
```
Hi @cmtice! I just came across this code and I don't think this API makes sense. Since `CastToBasicType` creates a new casted ValueObject I think this method should be `const` and not modify the error of the original value when the cast fails. I think it should return a ValueObjectConstResult with the error instead.

cmtice replied:
Ok, I have tried to do what I think you were asking me to do. Please see https://github.com/llvm/llvm-project/pull/117401
